test/common/websocket/codec_test.cc

```diff
@@ -0,0 +1,885 @@
+} // NOLINT(clang-analyzer-cplusplus.NewDeleteLeaks)
+
+TEST(WebSocketCodecTest, DecodeThreeFramesExceptPayload) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x01, 0x03, 0x48, 0x65, 0x6c});
+  Buffer::addSeq(buffer, {0x80, 0x02, 0x6c, 0x6f});
+  Buffer::addSeq(buffer, {0x81, 0x05, 0x48, 0x65, 0x6c, 0x6c, 0x6f});
+
+  Decoder decoder(0);
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(3, frames->size());
+  EXPECT_TRUE(
+      areFramesEqual(frames.value()[0], {false, kFrameOpcodeText, absl::nullopt, 3, nullptr}));
+  EXPECT_TRUE(areFramesEqual(frames.value()[1],
+                             {true, kFrameOpcodeContinuation, absl::nullopt, 2, nullptr}));
+  EXPECT_TRUE(
+      areFramesEqual(frames.value()[2], {true, kFrameOpcodeText, absl::nullopt, 5, nullptr}));
+}
+
+// A single-frame masked text message ("Hello")
+// 0x81 0x85 0x37 0xfa 0x21 0x3d 0x7f 0x9f 0x4d 0x51 0x58
+//           M1   M2   M3   M4   B1   B2   B3   B4   B5
+// fin - 1
+// opcode - 0x1 - text data
+// mask bit - 1
+// length - 5 bytes 0000101
+//
+// H - M1 XOR B1   01001000 = 00110111 xor 01111111
+// e - M2 XOR B2
+// l - M3 XOR B3
+// l - M4 XOR B4
+// o - M1 XOR B5
+//
+// j = i MOD 4
+// O(i) = M(j) XOR B(i)
+TEST(WebSocketCodecTest, DecodeSingleMaskedFrame) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x81, 0x85, 0x37, 0xfa, 0x21, 0x3d, 0x7f, 0x9f, 0x4d, 0x51, 0x58});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames->size());
+  EXPECT_TRUE(areFramesEqual(frames.value()[0], {true, kFrameOpcodeText, 0x37fa213d, 5,
+                                                 makeBuffer("\x7f\x9f\x4d\x51\x58")}));
+} // NOLINT(clang-analyzer-cplusplus.NewDeleteLeaks)
+
+// Unmasked Ping request and masked Ping response
+//
+// 0x89 0x05 0x48 0x65 0x6c 0x6c 0x6f (contains a body of "Hello",
+// but the contents of the body are arbitrary)
+//
+// 0x8a 0x85 0x37 0xfa 0x21 0x3d 0x7f 0x9f 0x4d 0x51 0x58
+// (contains a body of "Hello", matching the body of the ping)
+TEST(WebSocketCodecTest, DecodePingFrames) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x89, 0x05, 0x48, 0x65, 0x6c, 0x6c, 0x6f});
+  Buffer::addSeq(buffer, {0x8a, 0x85, 0x37, 0xfa, 0x21, 0x3d, 0x7f, 0x9f, 0x4d, 0x51, 0x58});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(2, frames->size());
+  EXPECT_TRUE(areFramesEqual(frames.value()[0],
+                             {true, kFrameOpcodePing, absl::nullopt, 5, makeBuffer("Hello")}));
+  EXPECT_TRUE(areFramesEqual(frames.value()[1], {true, kFrameOpcodePong, 0x37fa213d, 5,
+                                                 makeBuffer("\x7f\x9f\x4d\x51\x58")}));
+} // NOLINT(clang-analyzer-cplusplus.NewDeleteLeaks)
```

Comment:
May I ask why we need this?

Reply:
That is due to `makeBuffer` util function return a pointer. This is to avoided the linter error. I saw some other tests have added like this.